### ISSUE TITLE
High memory usage Environments

### ISSUE THREAD

**Issue by dmadeka** [ray-project/ray#3777]
Im working with an environment that has a very high memory usage. This usually prevents any sort of Async Sampling, since copies of the environment are very expensive. Is there any example of working with a high memory usage environment for ASync Sampling?

My understanding is that each state needs to be hashable, so it could be that because of the large state space - Im running quickly out of memory on instances with 500GB+ of RAM

**Comment by ericl** [comment]
Could you clarify what do you mean by high memory usage? Is it just that the observations are very large?

There are a number of ways to reduce memory usage from large observations. For example, you could reduce batch sizes, enable compression, or reduce replay buffer sizes. Certain algorithms like IMPALA also have options to reduce the amount if asynchronous buffering in the system that increases memory usage.

**Comment by dmadeka** [comment]
The observations are reasonably large. But they're all stored in memory, because a simulator (like most real world applications) isnt available :(

**Comment by ericl** [comment]
I see, so the issue is that you'd like to have multiple python class instances of your env share memory? Normally, parallelization with num_workers will launch a process per environment. This prevents class/static variable sharing.

An alternative is to use vectorization instead (num_envs_per_worker) which creates multiple env instances in the same python process enabling more efficient memory use. However the degree of parallelism is more limited here (you might not see any performance gains beyond 5-10 envs per worker). Or perhaps some mix of multiple workers and multiple envs per worker will be best.

**Comment by dmadeka** [comment]
So if I have a single plasma object store (is there an example of plasma stores for 3-4D numpy arrays?) on a single machine - can I have the workers index that (if the environment is defined that way?)

By the way, the accessibility of the Github issues/response/help is awesome! Thanks so very much

**Comment by ericl** [comment]
So if you `ray.put()` a numpy array, that array will be stored in shared memory, and when you ray.get() it on other tasks/actors, those workers just get a view of the shared array. The only trick is that you have to pass it into your RLlib env somehow, which currently requires a hack: https://github.com/ray-project/ray/issues/1453

Re: unhashable type -- what's the full error? If you implement a gym env all you need to do is set `num_envs_per_worker: N` to do vectorization, no need to manually implement.

**Comment by dmadeka** [comment]
Thats a good line to go with!

When I try to do `A3C` with `num_envs_per_worker`: 5, `num_workers`: 1. I get this error.

```python
~/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/ray/rllib/agents/a3c/a3c.py in _train(self)
     68         start = time.time()
     69         while time.time() - start < self.config["min_iter_time_s"]:
---> 70             self.optimizer.step()
     71         result = self.optimizer.collect_metrics(
     72             self.config["collect_metrics_timeout"])

~/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/ray/rllib/optimizers/async_gradients_optimizer.py in step(self)
     35             e.set_weights.remote(weights)
     36             future = e.compute_gradients.remote(e.sample.remote())
---> 37             pending_gradients[future] = e
     38             num_gradients += 1
     39 

TypeError: unhashable type: 'list'
```

**Comment by dmadeka** [comment]
future is a long list, with some odd shapes, and a `{'batch_count':1000}` (`batch_size` is set to `200`)

**Comment by ericl** [comment]
Are you running with ray.init(local_mode=True)? I don't think that really
works with A3C.

**Comment by dmadeka** [comment]
Yes, that and `huge_pages=False` - if i try to do it without that, initializing my object causes a `ConnectionResetError` because I guess the object cant be passed? If i turn local_mode off and load the data on the initialization of the object, it causes the memory to blow

So i guess i have to do the `put` `get` trick?

**Comment by dmadeka** [comment]
The local mode helps when you have a large instance, and want to run A3c with many threads and a single memory

**Comment by ericl** [comment]
In that case I would run A2C with num_workers=0. That's effectively what
you're getting in local mode anyways.

**Comment by dmadeka** [comment]
Also, quick question - does `PPO` support ASync sampling? I get an error because the default `observation_filter` becomes the `ConcurrentMeanStdFilter`

**Comment by dmadeka** [comment]
Wouldnt `A2C` with `num_workers=0` not allow many threads to generate samples?

**Comment by ericl** [comment]
RLlib doesn't use threading, only vectorization and multiple processes.
However this is no different from local mode (which is only for debugging
and has no parallelism at all).

Also there isn't async PPO yet.

**Comment by dmadeka** [comment]
Was the ASync PPO with IMPALA just merged? Would the master branch let me access it?

**Comment by goodcheer** [comment]
> So formally, every initialization of my environment class requires a lot of memory. There's a lot of data to be indexed, and only some parts of the environment are generated on the fly. So even before I start training, the initialization of A3C or any async sampler blows my memory. So, the question becomes - how do I store large collections of data in memory for many workers to index out of everytime a batch is computed.
> 
> Currently, all the data is stored in class variables, but Im guessing these are copied many times? Im just not sure where or what to override

Hi. I am experiencing the same issue as yours, and I would like to know how you figured out the problem. 

Furthermore, regarding the solution 

> An alternative is to use vectorization instead (num_envs_per_worker) which creates multiple env instances in the same python process enabling more efficient memory use. However the degree of parallelism is more limited here (you might not see any performance gains beyond 5-10 envs per worker). Or perhaps some mix of multiple workers and multiple envs per worker will be best.

Can you explain in more detail how 

> creates multiple env instances in the same python process enabling more efficient memory use

this mends the issue ? 
Thanks in advance.

**Comment by dmadeka** [comment]
@goodcheer I created a plasma store and indexed out of it. But its still pretty complex - some objects when deserialized create copies, which will blow the memory - https://github.com/ray-project/ray/issues/3881

But if you're data is in numpy arrays, the plasma store works really well to solve this problem
